server/service/src/sync/translation_and_integration.rs

```diff
@@ -0,0 +1,206 @@
+        let translators = all_translators();
+
+        for sync_record in sync_records {
+            // Try translate
+            let translation_result = match self.translate_sync_record(&sync_record, &translators) {
+                Ok(translation_result) => translation_result,
+                // Record error in sync buffer and in result, continue to next sync_record
+                Err(translation_error) => {
+                    self.sync_buffer
+                        .record_integration_error(&sync_record, &translation_error)?;
+                    result.insert_error(
+                        &sync_record.table_name,
+                        Some(&sync_record),
+                        None,
+                        translation_error,
+                    );
+                    // Next sync_record
+                    continue;
+                }
+            };
+
+            let integration_records = match translation_result {
+                Some(integration_records) => integration_records,
+                // Record translator not found error in sync buffer and in result, continue to next sync_record
+                None => {
+                    let error = anyhow::anyhow!("Translator for record not found");
+                    self.sync_buffer
+                        .record_integration_error(&sync_record, &error)?;
+                    result.insert_error(&sync_record.table_name, Some(&sync_record), None, error);
+                    // Next sync_record
+                    continue;
+                }
+            };
+
+            // Integrate
+            let integration_result = integration_records.integrate(self.connection);
+            match integration_result {
+                Ok(_) => {
+                    self.sync_buffer
+                        .record_successfull_integration(&sync_record)?;
+                    result.insert_success(&sync_record.table_name)
+                }
+                // Record database_error in sync buffer and in resul
+                Err(database_error) => {
+                    let error = anyhow::anyhow!("{:?}", database_error);
+                    self.sync_buffer
+                        .record_integration_error(&sync_record, &error)?;
+                    result.insert_error(
+                        &sync_record.table_name,
+                        Some(&sync_record),
+                        Some(&integration_records),
+                        error,
+                    );
+                }
+            }
+        }
+        Ok(result)
+    }
+}
+
+impl IntegrationRecords {
+    pub(crate) fn integrate(&self, connection: &StorageConnection) -> Result<(), RepositoryError> {
+        for upsert in self.upserts.iter() {
+            upsert.upsert(connection)?
+        }
+
+        Ok(())
+    }
+}
+
+impl PullUpsertRecord {
+    pub(crate) fn upsert(&self, con: &StorageConnection) -> Result<(), RepositoryError> {
```

Comment:
Bit unusual to have the DB mutation on the record. Could just be inlined in integrate()?

Reply:
I think it could have been in `integrate`, but it would obfuscate the transaction logic in `integrate` (or at least i would make it harder to digest).

Also if I was re-writing how we do database layer I would put base row repository on the record (and make it a trait), I think I talk about it in upstream PR description, same with inline_init, inline_edit, and the query repository would also go on the record type it returns. And would make a generic UpsertRecords and DeleteRecords struct that can take any row type, add it to collection and upsert/delete in correct order. (this can be used in test data, sync etc..)

Reply:
I would prefer a `upsert_record(con: &StorageConnection, record: &PullUpsertRecord)` method.

Regarding row struct, you would tight the DB logic quite strongly to our "domain" objects. Nice thing about repos is that you have it decoupled and its easier to swap our serialization logic....

Reply:
> Nice thing about repos is that you have it decoupled and its easier to swap our serialization logic

I've mentioned it upstream, i think it makes no difference really, swapping out implementation is the same as swapping a method

> I would prefer a upsert_record(con: &StorageConnection, record: &PullUpsertRecord) method.

I am on a different side of the argument here `upsert_record` -> the `record` part in that name implies you are identifying the method with some record, you can see which record it is by looking at the argument. To me, it's much better to 'identify' something with a type is to impl the behaviour on that type